File: packages/template-go/expr/lexer.go
// Package expr implements the expression lexer and parser of docs/spec/expressions.md.
package expr

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/polyspec/template/errs"
	"github.com/polyspec/template/lexer"
)

// TokenType is a token type name of CNF-13; CLOSE is the tag end.
type TokenType string

// Token is one expression token; Start and End are byte offsets.
type Token struct {
	Type    TokenType
	Value   string
	Start   int
	End     int
	Decoded string
}

var operators = []struct {
	text string
	typ  TokenType
}{
	{"===", "SEQ"}, {"!==", "SNE"}, {"...", "SPREAD"},
	{"==", "EQ"}, {"!=", "NE"}, {"<=", "LE"}, {">=", "GE"}, {"&&", "AND"}, {"||", "OR"},
	{"??", "COALESCE"}, {"?:", "ELVIS"}, {"=>", "ARROW"},
	{"(", "LPAREN"}, {")", "RPAREN"}, {"[", "LBRACKET"}, {"]", "RBRACKET"}, {",", "COMMA"},
	{"|", "PIPE"}, {"?", "QUESTION"}, {":", "COLON"}, {"+", "PLUS"}, {"-", "MINUS"},
	{"*", "STAR"}, {"/", "SLASH"}, {"%", "PERCENT"}, {"!", "BANG"}, {"<", "LT"}, {">", "GT"},
}

const expressionChars = "()[],|?:=>.+-*/%!<&'\""

var postfixEnd = map[TokenType]bool{"IDENT": true, "NUMBER": true, "STRING": true, "NULL": true, "TRUE": true, "FALSE": true, "RPAREN": true, "RBRACKET": true, "DOT_IDENT": true, "DOT_INDEX": true}

func isIdentStart(c byte) bool { return c >= 'A' && c <= 'Z' || c >= 'a' && c <= 'z' || c == '_' }
func isDigit(c byte) bool      { return c >= '0' && c <= '9' }
func isIdentPart(c byte) bool  { return isIdentStart(c) || isDigit(c) }
func isWhitespace(c byte) bool { return c == ' ' || c == '\t' || c == '\r' || c == '\n' }

// Options select the close delimiter of the enclosing tag.
type Options struct {
	Close      string // "" for a bare expression
	CloseCount int
	OpenIndex  int // -1 for a bare expression
}

// Lexer produces expression tokens.
type Lexer struct {
	source       *lexer.Source
	options      Options
	index        int
	previous     *Token
	lookahead    *Token
	closeExpr    bool
	nestingDepth int
}

// NewLexer creates a lexer starting at a byte offset.
func NewLexer(source *lexer.Source, start int, options Options) *Lexer {
	return &Lexer{source: source, options: options, index: start, closeExpr: options.Close != "" && strings.Contains(expressionChars, options.Close)}
}

// Error creates an error located in the source.
func (l *Lexer) Error(code errs.Code, start, end int, message string) *errs.Error {
	return l.source.Error(code, start, end, message)
}

// Peek returns the next token without consuming it.
func (l *Lexer) Peek() (*Token, error) {
	if l.lookahead == nil {
		tok, err := l.read()
		if err != nil {
			return nil, err
		}
		l.lookahead = tok
	}
	return l.lookahead, nil
}

// Next consumes and returns the next token.
func (l *Lexer) Next() (*Token, error) {
	tok, err := l.Peek()
	if err != nil {
		return nil, err
	}
	l.lookahead = nil
	l.previous = tok
	if tok.Type == "LPAREN" || tok.Type == "LBRACKET" {
		l.nestingDepth++
	}
	if tok.Type == "RPAREN" || tok.Type == "RBRACKET" {
		l.nestingDepth--
	}
	return tok, nil
}

// ConsumedEnd is the byte offset after the last consumed token.
func (l *Lexer) ConsumedEnd() int {
	if l.previous != nil {
		return l.previous.End
	}
	return l.index
}

func (l *Lexer) read() (*Token, error) {
	text := l.source.Text
	for l.index < len(text) && isWhitespace(text[l.index]) {
		l.index++
	}
	start := l.index
	if start >= len(text) {
		return &Token{Type: "EOF", Start: start, End: start}, nil
	}
	if l.options.Close != "" && (l.previous != nil && postfixEnd[l.previous.Type] && l.nestingDepth == 0 || !l.closeExpr) {
		sequence := strings.Repeat(l.options.Close, l.options.CloseCount)
		if strings.HasPrefix(text[start:], sequence) {
			l.index = start + len(sequence)
			return &Token{Type: "CLOSE", Value: sequence, Start: start, End: l.index}, nil
		}
	}
	c := text[start]
	switch {
	case isIdentStart(c):
		end := start + 1
		for end < len(text) && isIdentPart(text[end]) {
			end++
		}
		l.index = end
		value := text[start:end]
		typ := TokenType("IDENT")
		switch value {
		case "null":
			typ = "NULL"
		case "true":
			typ = "TRUE"
		case "false":
			typ = "FALSE"
		case "in":
			typ = "IN"
		}
		return &Token{Type: typ, Value: value, Start: start, End: end}, nil
	case isDigit(c):
		return l.readNumber(start)
	case c == '"' || c == '\'':
		literal, err := LexStringLiteral(l.source, start)
		if err != nil {
			return nil, err
		}
		l.index = literal.End
		return &Token{Type: "STRING", Value: text[start:literal.End], Start: start, End: literal.End, Decoded: literal.Decoded}, nil
	case c == '.':
		return l.readDot(start)
	}
	for _, op := range operators {
		if strings.HasPrefix(text[start:], op.text) {
			l.index = start + len(op.text)
			return &Token{Type: op.typ, Value: op.text, Start: start, End: l.index}, nil
		}
	}
	return nil, l.Error(errs.ParseUnexpectedToken, start, start+1, fmt.Sprintf("unexpected character %q", text[start:start+1]))
}

func (l *Lexer) readNumber(start int) (*Token, error) {
	text := l.source.Text
	end := start
	for end < len(text) && isDigit(text[end]) {
		end++
	}
	if end+1 < len(text) && text[end] == '.' && isDigit(text[end+1]) {
		end++
		for end < len(text) && isDigit(text[end]) {
			end++
		}
	}
	if end < len(text) && (text[end] == 'e' || text[end] == 'E') {
		cursor := end + 1
		if cursor < len(text) && (text[cursor] == '+' || text[cursor] == '-') {
			cursor++
		}
		if cursor < len(text) && isDigit(text[cursor]) {
			for cursor < len(text) && isDigit(text[cursor]) {
				cursor++
			}
			end = cursor
		} else {
			return nil, l.Error(errs.ParseInvalidNumber, start, cursor, fmt.Sprintf("invalid number %q", text[start:cursor]))
		}
	}
	closeAtEnd := l.options.Close != "" && l.nestingDepth == 0 && strings.HasPrefix(text[end:], l.options.Close)
	if end < len(text) && (isIdentPart(text[end]) || (text[end] == '.' && !closeAtEnd)) {
		cursor := end + 1
		for cursor < len(text) && (isIdentPart(text[cursor]) || text[cursor] == '.') {
			cursor++
		}
		return nil, l.Error(errs.ParseInvalidNumber, start, cursor, fmt.Sprintf("invalid number %q", text[start:cursor]))
	}
	l.index = end
	return &Token{Type: "NUMBER", Value: text[start:end], Start: start, End: end}, nil
}

func (l *Lexer) readDot(start int) (*Token, error) {
	text := l.source.Text
	if strings.HasPrefix(text[start:], "...") {
		l.index = start + 3
		return &Token{Type: "SPREAD", Value: "...", Start: start, End: l.index}, nil
	}
	var next byte
	if start+1 < len(text) {
		next = text[start+1]
	}
	adjacent := l.previous != nil && l.previous.End == start && postfixEnd[l.previous.Type]
	if adjacent && isIdentStart(next) {
		end := start + 2
		for end < len(text) && isIdentPart(text[end]) {
			end++
		}
		l.index = end
		return &Token{Type: "DOT_IDENT", Value: text[start:end], Start: start, End: end}, nil
	}
	if adjacent && isDigit(next) {
		end := start + 2
		for end < len(text) && isDigit(text[end]) {
			end++
		}
		l.index = end
		return &Token{Type: "DOT_INDEX", Value: text[start:end], Start: start, End: end}, nil
	}
	if isDigit(next) {
		end := start + 1
		for end < len(text) && (isIdentPart(text[end]) || text[end] == '.') {
			end++
		}
		return nil, l.Error(errs.ParseInvalidNumber, start, end, fmt.Sprintf("invalid number %q", text[start:end]))
	}
	return nil, l.Error(errs.ParseUnexpectedToken, start, start+1, "unexpected \".\"")
}

// StringLiteral is a decoded string literal.
type StringLiteral struct {
	Decoded string
	End     int
}

// LexStringLiteral reads a string literal starting at the quote at start (EXP-3).
func LexStringLiteral(source *lexer.Source, start int) (StringLiteral, error) {
	text := source.Text
	quote := text[start]
	var decoded strings.Builder
	index := start + 1
	for {
		if index >= len(text) {
			return StringLiteral{}, source.Error(errs.ParseUnterminatedString, start, start+1, "string literal is not terminated")
		}
		c := text[index]
		if c == quote {
			return StringLiteral{Decoded: decoded.String(), End: index + 1}, nil
		}
		if c != '\\' {
			decoded.WriteByte(c)
			index++
			continue
		}
		var escape byte
		if index+1 < len(text) {
			escape = text[index+1]
		}
		switch escape {
		case '\\':
			decoded.WriteByte('\\')
			index += 2
		case '\'':
			decoded.WriteByte('\'')
			index += 2
		case '"':
			decoded.WriteByte('"')
			index += 2
		case 'n':
			decoded.WriteByte('\n')
			index += 2
		case 'r':
			decoded.WriteByte('\r')
			index += 2
		case 't':
			decoded.WriteByte('\t')
			index += 2
		case 'u':
			if index+6 > len(text) || !isHex(text[index+2:index+6]) {
				return StringLiteral{}, source.Error(errs.ParseInvalidEscape, index, index+2, "invalid escape sequence")
			}
			code, _ := strconv.ParseUint(text[index+2:index+6], 16, 32)
			index += 6
			if code >= 0xd800 && code <= 0xdbff && index+6 <= len(text) && text[index] == '\\' && text[index+1] == 'u' && isHex(text[index+2:index+6]) {
				low, _ := strconv.ParseUint(text[index+2:index+6], 16, 32)
				if low >= 0xdc00 && low <= 0xdfff {
					code = 0x10000 + (code-0xd800)<<10 + (low - 0xdc00)
					index += 6
				}
			}
			decoded.WriteRune(rune(code))
		default:
			return StringLiteral{}, source.Error(errs.ParseInvalidEscape, index, index+2, "invalid escape sequence")
		}
	}
}

func isHex(s string) bool {
	if len(s) != 4 {
		return false
	}
	for i := 0; i < len(s); i++ {
		c := s[i]
		if !(isDigit(c) || c >= 'a' && c <= 'f' || c >= 'A' && c <= 'F') {
			return false
		}
	}
	return true
}

File: packages/template-ts/src/expr/lexer.ts
// Expression tokens as defined in EXP-1 to EXP-6 and CNF-13.
import { errorAt, type ErrorCode, type TemplateError } from '../errors.js';
import type { Source } from '../source.js';

export type TokenType =
  | 'IDENT' | 'NUMBER' | 'STRING' | 'DOT_IDENT' | 'DOT_INDEX'
  | 'LPAREN' | 'RPAREN' | 'LBRACKET' | 'RBRACKET' | 'COMMA' | 'PIPE'
  | 'QUESTION' | 'COLON' | 'ELVIS' | 'COALESCE' | 'ARROW' | 'SPREAD'
  | 'PLUS' | 'MINUS' | 'STAR' | 'SLASH' | 'PERCENT' | 'BANG'
  | 'EQ' | 'NE' | 'SEQ' | 'SNE' | 'LT' | 'GT' | 'LE' | 'GE' | 'AND' | 'OR' | 'IN'
  | 'NULL' | 'TRUE' | 'FALSE' | 'EOF'
  | 'CLOSE';

export interface Token {
  type: TokenType;
  value: string;
  // String indexes into the source text.
  start: number;
  end: number;
  // Decoded string value for STRING tokens.
  decoded?: string;
}

const OPERATORS: [string, TokenType][] = [
  ['===', 'SEQ'], ['!==', 'SNE'], ['...', 'SPREAD'],
  ['==', 'EQ'], ['!=', 'NE'], ['<=', 'LE'], ['>=', 'GE'], ['&&', 'AND'], ['||', 'OR'],
  ['??', 'COALESCE'], ['?:', 'ELVIS'], ['=>', 'ARROW'],
  ['(', 'LPAREN'], [')', 'RPAREN'], ['[', 'LBRACKET'], [']', 'RBRACKET'], [',', 'COMMA'],
  ['|', 'PIPE'], ['?', 'QUESTION'], [':', 'COLON'], ['+', 'PLUS'], ['-', 'MINUS'],
  ['*', 'STAR'], ['/', 'SLASH'], ['%', 'PERCENT'], ['!', 'BANG'], ['<', 'LT'], ['>', 'GT'],
];

// Characters that start an expression token; a close delimiter among them is lexed as its token.
const EXPRESSION_CHARS = new Set('()[],|?:=>.+-*/%!<&\'"'.split(''));

const POSTFIX_END: ReadonlySet<TokenType> = new Set(['IDENT', 'RPAREN', 'RBRACKET', 'DOT_IDENT', 'DOT_INDEX']);

function isIdentStart(code: number): boolean {
  return (code >= 0x41 && code <= 0x5a) || (code >= 0x61 && code <= 0x7a) || code === 0x5f;
}

function isIdentPart(code: number): boolean {
  return isIdentStart(code) || isDigit(code);
}

function isDigit(code: number): boolean {
  return code >= 0x30 && code <= 0x39;
}

function isWhitespace(code: number): boolean {
  return code === 0x20 || code === 0x09 || code === 0x0d || code === 0x0a;
}

export interface LexerOptions {
  // Close delimiter character and how many times it is repeated to end the tag; null for a bare expression.
  close: string | null;
  closeCount: number;
  // String index of the tag start, used for E_PARSE_UNTERMINATED_TAG; null for a bare expression.
  openIndex: number | null;
}

export interface StringLiteral {
  decoded: string;
  end: number;
}

// Reads a string literal starting at the quote at `start` (EXP-3).
export function lexStringLiteral(source: Source, start: number, template: string): StringLiteral {
  const text = source.text;
  const quote = text[start] as string;
  const fail = (code: ErrorCode, from: number, to: number, message: string): TemplateError =>
    errorAt(code, template, source.lines, source.span(from, to), message);
  let decoded = '';
  let index = start + 1;
  for (;;) {
    if (index >= text.length) throw fail('E_PARSE_UNTERMINATED_STRING', start, start + 1, 'string literal is not terminated');
    const char = text[index] as string;
    if (char === quote) return { decoded, end: index + 1 };
    if (char !== '\\') {
      decoded += char;
      index++;
      continue;
    }
    const escape = text[index + 1];
    switch (escape) {
      case '\\': decoded += '\\'; index += 2; break;
      case "'": decoded += "'"; index += 2; break;
      case '"': decoded += '"'; index += 2; break;
      case 'n': decoded += '\n'; index += 2; break;
      case 'r': decoded += '\r'; index += 2; break;
      case 't': decoded += '\t'; index += 2; break;
      case 'u': {
        const hex = text.slice(index + 2, index + 6);
        if (!/^[0-9a-fA-F]{4}$/.test(hex)) throw fail('E_PARSE_INVALID_ESCAPE', index, index + 2, 'invalid escape sequence');
        decoded += String.fromCharCode(parseInt(hex, 16));
        index += 6;
        break;
      }
      default:
        throw fail('E_PARSE_INVALID_ESCAPE', index, index + 2, 'invalid escape sequence');
    }
  }
}

export class ExpressionLexer {
  private index: number;
  private previous: Token | null = null;
  private lookahead: Token | null = null;
  private readonly closeIsExpressionChar: boolean;

  constructor(
    private readonly source: Source,
    start: number,
    private readonly options: LexerOptions,
    private readonly template: string,
  ) {
    this.index = start;
    this.closeIsExpressionChar = options.close !== null && EXPRESSION_CHARS.has(options.close);
  }

  get position(): number {
    return this.lookahead ? this.lookahead.start : this.index;
  }

  error(code: ErrorCode, start: number, end: number, message: string): TemplateError {
    return errorAt(code, this.template, this.source.lines, this.source.span(start, end), message);
  }

  peek(): Token {
    if (!this.lookahead) this.lookahead = this.read();
    return this.lookahead;
  }

  next(): Token {
    const token = this.peek();
    this.lookahead = null;
    this.previous = token;
    return token;
  }

  // Byte offset after the last consumed token.
  get consumedEnd(): number {
    return this.previous ? this.previous.end : this.index;
  }

  private read(): Token {
    const text = this.source.text;
    while (this.index < text.length && isWhitespace(text.charCodeAt(this.index))) this.index++;
    const start = this.index;
    if (start >= text.length) return { type: 'EOF', value: '', start, end: start };

    const close = this.options.close;
    if (close !== null && !this.closeIsExpressionChar) {
      const sequence = close.repeat(this.options.closeCount);
      if (text.startsWith(sequence, start)) {
        this.index = start + sequence.length;
        return { type: 'CLOSE', value: sequence, start, end: this.index };
      }
    }

    const code = text.charCodeAt(start);
    if (isIdentStart(code)) {
      let end = start + 1;
      while (end < text.length && isIdentPart(text.charCodeAt(end))) end++;
      this.index = end;
      const value = text.slice(start, end);
      const type: TokenType =
        value === 'null' ? 'NULL' : value === 'true' ? 'TRUE' : value === 'false' ? 'FALSE' : value === 'in' ? 'IN' : 'IDENT';
      return { type, value, start, end };
    }
    if (isDigit(code)) return this.readNumber(start);
    if (code === 0x22 || code === 0x27) return this.readString(start);
    if (code === 0x2e) return this.readDot(start);

    for (const [operator, type] of OPERATORS) {
      if (text.startsWith(operator, start)) {
        this.index = start + operator.length;
        return { type, value: operator, start, end: this.index };
      }
    }
    throw this.error('E_PARSE_UNEXPECTED_TOKEN', start, start + 1, `unexpected character ${JSON.stringify(text[start])}`);
  }

  private readNumber(start: number): Token {
    const text = this.source.text;
    let end = start;
    while (end < text.length && isDigit(text.charCodeAt(end))) end++;
    if (text[end] === '.' && end + 1 < text.length && isDigit(text.charCodeAt(end + 1))) {
      end++;
      while (end < text.length && isDigit(text.charCodeAt(end))) end++;
    }
    if (text[end] === 'e' || text[end] === 'E') {
      let cursor = end + 1;
      if (text[cursor] === '+' || text[cursor] === '-') cursor++;
      if (cursor < text.length && isDigit(text.charCodeAt(cursor))) {
        while (cursor < text.length && isDigit(text.charCodeAt(cursor))) cursor++;
        end = cursor;
      } else {
        throw this.error('E_PARSE_INVALID_NUMBER', start, cursor, `invalid number ${JSON.stringify(text.slice(start, cursor))}`);
      }
    }
    const following = text.charCodeAt(end);
    if (end < text.length && (isIdentPart(following) || following === 0x2e)) {
      let cursor = end + 1;
      while (cursor < text.length && (isIdentPart(text.charCodeAt(cursor)) || text[cursor] === '.')) cursor++;
      throw this.error('E_PARSE_INVALID_NUMBER', start, cursor, `invalid number ${JSON.stringify(text.slice(start, cursor))}`);
    }
    this.index = end;
    return { type: 'NUMBER', value: text.slice(start, end), start, end };
  }

  private readDot(start: number): Token {
    const text = this.source.text;
    if (text.startsWith('...', start)) {
      this.index = start + 3;
      return { type: 'SPREAD', value: '...', start, end: this.index };
    }
    const nextCode = text.charCodeAt(start + 1);
    const adjacent = this.previous !== null && this.previous.end === start && POSTFIX_END.has(this.previous.type);
    if (adjacent && isIdentStart(nextCode)) {
      let end = start + 2;
      while (end < text.length && isIdentPart(text.charCodeAt(end))) end++;
      this.index = end;
      return { type: 'DOT_IDENT', value: text.slice(start, end), start, end };
    }
    if (adjacent && isDigit(nextCode)) {
      let end = start + 2;
      while (end < text.length && isDigit(text.charCodeAt(end))) end++;
      this.index = end;
      return { type: 'DOT_INDEX', value: text.slice(start, end), start, end };
    }
    if (isDigit(nextCode)) {
      let end = start + 1;
      while (end < text.length && (isIdentPart(text.charCodeAt(end)) || text[end] === '.')) end++;
      throw this.error('E_PARSE_INVALID_NUMBER', start, end, `invalid number ${JSON.stringify(text.slice(start, end))}`);
    }
    throw this.error('E_PARSE_UNEXPECTED_TOKEN', start, start + 1, 'unexpected "."');
  }

  private readString(start: number): Token {
    const literal = lexStringLiteral(this.source, start, this.template);
    this.index = literal.end;
    return { type: 'STRING', value: this.source.text.slice(start, literal.end), start, end: literal.end, decoded: literal.decoded };
  }
}
